### PR TITLE
move ics4wrapper from keeper to ibc module

### PR DESCRIPTION
**Pull request by GNaD13**
## Description

closes: #3371 

### Commit Message / Changelog Entry

```bash
imp(api): remove ICS4Wrapper from keepers and pass it as argument to middleware constructor
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by faddat** [comment]
@GNaD13 please provide a commit message for this pr and btw, great work!

**Comment by GNaD13** [comment]
> @GNaD13 please provide a commit message for this pr and btw, great work!

Got it sir

**Comment by faddat** [comment]
nice :D.  after I merged main, it blew up your work a bit I think, sorry

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go [hunk 1]
```diff
@@ -23,13 +23,15 @@ var _ porttypes.Middleware = (*IBCMiddleware)(nil)
 type IBCMiddleware struct {
 	app    porttypes.IBCModule
 	keeper keeper.Keeper
+	next   porttypes.ICS4Wrapper
```
[the remark this answers is not in the document]
Comment (GNaD13) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:26]: Hmmmm, I'm not really get it. Can you explain more detail please?

[the remark this answers is not in the document]
Comment (DimitrisJim) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:26]: Looking a bit more into it, don't think you can actually do what I hoped (the idea was to remove `ics4Wrapper` from `controller.Keeper`) due to `SendTx`. Sorry for any confusion!

File: testing/simapp/app.go [hunk 2]
```diff
@@ -483,17 +485,16 @@ func NewSimApp(
 
 	// initialize ICA module with mock module as the authentication module on the controller side
 	var icaControllerStack porttypes.IBCModule
-	icaControllerStack = ibcmock.NewIBCModule(&mockModule, ibcmock.NewIBCApp("", scopedICAMockKeeper))
-	app.ICAAuthModule = icaControllerStack.(ibcmock.IBCModule)
```
Comment (crodriguezvega) [inline testing/simapp/app.go:487]: Why was this changed?

File: testing/simapp/app.go [hunk 3]
```diff
@@ -483,17 +485,16 @@ func NewSimApp(
...
-	icaControllerStack = icacontroller.NewIBCMiddleware(icaControllerStack, app.ICAControllerKeeper)
-	icaControllerStack = ibcfee.NewIBCMiddleware(icaControllerStack, app.IBCFeeKeeper)
-
+	ibcMockModule := ibcmock.NewIBCModule(&mockModule, ibcmock.NewIBCApp("", scopedICAMockKeeper))
+	app.ICAAuthModule = ibcMockModule
+	icaControllerMiddleware := icacontroller.NewIBCMiddleware(ibcMockModule, app.ICAControllerKeeper, app.IBCKeeper.ChannelKeeper)
+	icaControllerStack = ibcfee.NewIBCMiddleware(icaControllerMiddleware, app.IBCFeeKeeper, icaControllerMiddleware)
```
Comment (crodriguezvega) [inline testing/simapp/app.go:491]: We are passing here the `icaControllerMiddleware` twice, one for `app` and one for `next`. Do we like this pattern?

**Review by crodriguezvega — COMMENTED** [review]
Thank you very much for your contribution, @GNaD13. And sorry for the late response from our side. I left a few comments, but I would like to hear @colin-axner or @damiannolan opinion to confirm if they like this pattern.
